crates/wasi-experimental-http-wasmtime/src/lib.rs

```diff
@@ -3,190 +3,374 @@ use bytes::Bytes;
 use futures::executor::block_on;
 use http::{HeaderMap, HeaderValue};
 use reqwest::{Client, Method};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
 use std::str::FromStr;
 use tokio::runtime::Handle;
 use url::Url;
 use wasmtime::*;
 
-const ALLOC_FN: &str = "alloc";
 const MEMORY: &str = "memory";
 
-pub fn link_http(linker: &mut Linker, allowed_hosts: Option<Vec<String>>) -> Result<(), Error> {
-    linker.func(
-        "wasi_experimental_http",
-        "req",
-        move |caller: Caller<'_>,
-              url_ptr: u32,
-              url_len: u32,
-              method_ptr: u32,
-              method_len: u32,
-              req_body: u32,
-              req_body_len: u32,
-              headers_ptr: u32,
-              headers_len: u32,
-              body_res_ptr: u32,
-              body_written_ptr: u32,
-              headers_res_ptr: u32,
-              headers_written_ptr: u32,
-              status_code_ptr: u32,
-              err_ptr: u32,
-              err_written_ptr: u32|
-              -> u32 {
-            let span = tracing::trace_span!("req");
-            let _enter = span.enter();
-            // Get the module's memory and allocation function.
-            // If either is not found, the runtime cannot write any response
-            // data, so the execution cannot continue.
-            tracing::trace!(export = MEMORY, "getting export");
-            let memory = match caller.get_export(MEMORY) {
-                Some(Extern::Memory(mem)) => mem,
-                _ => {
-                    return err(
-                        "cannot find memory".to_string(),
-                        None,
-                        None,
-                        err_ptr,
-                        err_written_ptr,
-                        1,
-                    )
+pub type WasiHandle = u32;
+
+struct Body {
+    bytes: Vec<u8>,
+    pos: usize,
+}
+
+struct Response {
+    headers: HeaderMap,
+    body: Body,
+}
+
+#[derive(Default)]
+pub struct State {
+    responses: HashMap<WasiHandle, Response>,
```

Comment:
If I understand correctly, when reading a response, a guest module will use the `WasiHandle` returned by the runtime.
But since the state has responses as a hash map of integers and responses, and the handle integers are incremented before adding a new response to the state, there is nothing stopping a guest module from incrementing or decrementing their handle and read a different guest's response.
(or even read the responses from _all_ guests running in the runtime.)

Does the handle need to be an integer?
Could we have a non-sequential ID instead?

Reply:
Handles are instance-specific. Each instance has their own handle space. Similar to descriptors in WASI core. So a module cannot access another one's handles.

The current development code of `wasmtime` also includes a handle manager, so we can switch to it once the next version is released. The handle manager is just a hash map, but it exposes a common type that WASI extensions can use.

Handles are defined as `handle` in WITX, which is just an alias for `u32`.

Reply:
In the WASI crypto extensions, handles are also incrementing, but the top bits encode a type. So that the type of a handle can be checked without even having to do a lookup.

But here, we only have one type, so, it's probably not worth it yet. Maybe once we support streaming for outgoing bodies, though.

Reply:
Ah, you are right.
A new state is created whenever instantiating a new module.

In this case, an integer is perfectly fine, and we can switch to Wasmtime's manager once that is released.